### PR TITLE
Further settings work

### PR DESCRIPTION
**Pull request by BertZZ**
Neatens up the settings page a little to make it more user friendly. 
Adds validations for the fields that are likely to stay on this form. 
Tests to follow.

### CODE REVIEW

File: app/models/concerns/hyku_addons/account_behavior.rb [hunk 1]
```diff
@@ -83,5 +102,12 @@ def remove_settings_hash_key_with_nil_value
           settings[key].delete_if { |_hash_key, value| value.blank? } if settings[key].class == Hash
         end
       end
+
+      def map_array_fields
```
Comment (cjcolvar) [inline app/models/concerns/hyku_addons/account_behavior.rb:106]: This seems more like a controller or form concern to me than a model concern.  What do you think?

File: app/models/concerns/hyku_addons/account_behavior.rb [hunk 2]
```diff
@@ -31,6 +37,19 @@ def datacite_endpoint
 
     private
 
+      def validate_email_format
```
Comment (cjcolvar) [inline app/models/concerns/hyku_addons/account_behavior.rb:40]: It's not clear to me if this method is intended for checking if `email_format` is in the proper format or is intended to do data manipulation.  Same for `validate_contact_emails`.  I would expect validation methods to only be checking for proper format and data manipulation to be happening in the controller or form or setter methods.

File: app/models/concerns/hyku_addons/account_behavior.rb [hunk 3]
```diff
@@ -3,7 +3,9 @@
 module HykuAddons
   module AccountBehavior
     extend ActiveSupport::Concern
+    include ActiveModel::Validations
     included do
+      email_validation_format = URI::MailTo::EMAIL_REGEXP
```
Comment (cjcolvar) [inline app/models/concerns/hyku_addons/account_behavior.rb:8]: Do we need to store this in a variable?  It feels like it should either be a constant or skipped and use `URI::MailTo::EMAIL_REGEXP` directly in the `validates` below.

**Review by cjcolvar — COMMENTED** [review]
Looks good. I just had a few questions/suggestions.

File: app/models/concerns/hyku_addons/account_behavior.rb (hunk 2 above)
Comment (okeen) [reply app/models/concerns/hyku_addons/account_behavior.rb:40]: There is a lot of validation logic into the Account model that is not intrinsic to the Account itself, but dependent on the context (for example, should an account always validate the settings email format, or rather only do it when you are updating settings?). 

It looks like we could extract all the settings functionality into a Form object as Chris says. https://codeclimate.com/blog/7-ways-to-decompose-fat-activerecord-models/

File: app/models/concerns/hyku_addons/account_behavior.rb [hunk 4]
```diff
@@ -22,7 +24,11 @@ module AccountBehavior
       after_initialize :set_jsonb_help_texts_default_keys, :set_jsonb_work_unwanted_fields_default_keys
       after_initialize :set_jsonb_required_json_property_default_keys, :set_jsonb_html_required_default_keys
       after_initialize :set_jsonb_metadata_labels_default_keys, :set_jsonb_licence_list_default_keys
+      before_validation :map_array_fields
       before_save :remove_settings_hash_key_with_nil_value
+      validates :gtm_id, :demo_gtm_id, format: { with: /GTM-[A-Z0-9]{4,7}/, message: "Invalid GTM ID" }, allow_blank: true
```
Comment (okeen) [inline app/models/concerns/hyku_addons/account_behavior.rb:29]: I think putting this into Rails' Validator clases would make it easier to read, maintain and refactor in the future:

`validates :gtm_id, :demo_gtm_id, gtm: true`

https://api.rubyonrails.org/classes/ActiveModel/Validator.html

File: app/models/concerns/hyku_addons/account_behavior.rb [hunk 5]
```diff
@@ -31,6 +37,19 @@ def datacite_endpoint
...
     private
 
+      def validate_email_format
+        return unless settings['email_format'].present?
+        settings['email_format'] = settings['email_format'].grep(/@\S*\.\S*/)
+      end
+
+      def validate_contact_emails
+        email_validation_format = URI::MailTo::EMAIL_REGEXP
+        ['weekly_email_list', 'monthly_email_list', 'yearly_email_list'].each do |key|
```
Comment (okeen) [inline app/models/concerns/hyku_addons/account_behavior.rb:47]: Should we add a class property + accessor for the weekly, monthly and yearly list array?

@@AVAILABLE_EMAIL_PERIODS = %w(week month year).map{ |period| "#{period}ly_email_list" }
cattr_reader :AVAILABLE_EMAIL_PERIODS

We could even set the periods on the settings.yml file

File: app/views/hyku_addons/account_settings/index.html.erb [hunk 6]
```diff
@@ -19,36 +19,34 @@
 
     <% ['contact_email', 'gtm_id', 'demo_gtm_id', 'live', 'institutional_relationship', 'creator_fields', 'contributor_fields'].each do |fname| %>
```
Comment (okeen) [inline app/views/hyku_addons/account_settings/index.html.erb:20]: We could extract this into a rails helper that accepts a block. That would improve readability which is extremely important on the views IMO.

Something like:
```

def for_available_setting_field
  Account.contact_settings_fields.each do |field|
    capture(yield(field))
  end
end
```

And then use it as:

```
<% for_available_setting_field do |field | %>
....

```

File: app/views/hyku_addons/account_settings/index.html.erb [hunk 7]
```diff
@@ -19,36 +19,34 @@
...
       <tr>
-        <td id= "link-<%= fname %>"><b><%= fname %>:</b> <%= @account && @account.settings && @account.settings[fname] %> </td> <br>
-        <td><%= link_to "Edit #{fname}", hyku_addons.edit_admin_account_setting_path(id: @account, partial_name: 'render_text_fields', field_name: fname), class: "btn btn-xs btn-default",  remote: true %></td>
+        <td id= "link-<%= fname %>"><b><%= fname.humanize %>:</b> <%= @account && @account.settings && @account.settings[fname] %> </td> <br>
```
Comment (okeen) [inline app/views/hyku_addons/account_settings/index.html.erb:22]: I think we can extract this into a helper method + presenters.
Something like:

<td><%= link_to_edit_account_settings %></td>

https://www.rubyguides.com/2019/09/rails-patterns-presenter-service/

**Review by okeen — COMMENTED** [review]
I have attached a few comments inline. Sorry if the formatting is not the best :(

My very first GH review ever :)

File: app/models/concerns/hyku_addons/account_behavior.rb (hunk 3 above)
Comment (BertZZ) [reply app/models/concerns/hyku_addons/account_behavior.rb:8]: Yeah you are probably right here.

File: app/models/concerns/hyku_addons/account_behavior.rb (hunk 2 above)
Comment (BertZZ) [reply app/models/concerns/hyku_addons/account_behavior.rb:40]: I see what you are saying @cjcolvar let me try something else

File: app/models/concerns/hyku_addons/account_behavior.rb (hunk 1 above)
Comment (BertZZ) [reply app/models/concerns/hyku_addons/account_behavior.rb:106]: Yeah let me try moving it to the controller

File: app/views/hyku_addons/account_settings/index.html.erb (hunk 6 above)
Comment (BertZZ) [reply app/views/hyku_addons/account_settings/index.html.erb:20]: I agree but I'm not sure it's something I want to spend time on now

**Review by cjcolvar — APPROVED** [review]
Looks good! :tada:

**Comment by cjcolvar** [comment]
I'm not really sure what's up with the total coverage dropping when the diff coverage is 100%.

**Pull request merged**